_posts/en/pages/2016-01-18-segwit-wallet-dev.md

```diff
@@ -8,129 +8,165 @@ lang: en
+* Examples of the transaction serialization could be found under the example section of BIP143. Wallet developers may use the examples to test if their implementations correctly parse the new serialization format.
+
+#### Transaction ID
+
+* Under segwit, each transaction will have 2 IDs.
+* Definition of <code>txid</code> remains unchanged: the double SHA256 of the original serialization format.
+* A new <code>wtxid</code> is defined, which is the double SHA256 of the new serialization format with witness data.
+* If a transaction does not have any witness data, its <code>wtxid</code> is same as the <code>txid</code>.
+* The <code>txid</code> remains the primary identifier of a transaction:
+    * It MUST be used in the <code>txin</code> when referring to a previous output.
+    * If a wallet or service is currently using <code>txin</code> to identify transactions, it is expected to use the same after upgrade.
+
+#### Signature Generation and Verification for P2SH-P2WPKH
+
+* For spending of non-segwit UTXO, the signature generation algorithm is unchanged.
+* For spending of P2SH-P2WPKH:
+    * The <code>scriptSig</code> MUST ONLY contain a push of the <code>redeemScript</code>
+    * The corresponding witness field MUST contain exactly 2 items, a signature followed by the public key
+    * There is a new signature generation algorithm described in BIP143 for segwit scripts. Developers should follow the instructions carefully, and make use of the P2SH-P2WPKH example in BIP143 to make sure they are able to reproduce the <code>sighash</code> [https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki#P2SHP2WPKH].
+    * The BIP143 signature generating algorithm covers the value of the input being spent, which simplifies the design of air-gapped light-weight wallet.
+    * Please note that for a P2SH-P2WPKH or native P2WPKH (not required for a basic segwit wallet), the <code>scriptCode</code> is always 26 bytes including the leading size byte, as <code>0x1976a914{20-byte keyhash}88ac</code>, NOT the <code>redeemScript</code> nor <code>scriptPubKey</code>.
+    
+#### Network Services (optional)
+
+* Network services is needed if the wallet would send and receive transactions through the peer-to-peer network
+* Segregated witness capable nodes will advertise that they can provide witnesses through the service bit: <code>NODE_WITNESS = (1 << 3)</code>
+* Transactions without any data (and therefore serialized in original format) may be sent to nodes with or without <code>NODE_WITNESS</code> support
+* Transactions with spending of segwit UTXO (and therefore is serialized in the new format) MUST ONLY be sent to nodes with <code>NODE_WITNESS</code> support
+* Transactions with spending of segwit UTXO but with witness data stripped (and therefore serialized in original format) may be sent to nodes without <code>NODE_WITNESS</code>. Such transactions, however, are invalid after activation of segwit and would not be accepted in a block.
```

Comment:
I'd strip this since those transactions are non-standard anyways and won't successfully propagate.